vllm/compilation/backends.py

```diff
@@ -164,80 +166,212 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
-                                             additional_inductor_config)
-
-    # TODO: Dynamo does not pass all dynamic shapes.
-    # Need to investigate why. It works now because all the dynamic
-    # shapes have the same value, and either of them can be used.
-    sym_shape_indices = [
-        i for i, x in enumerate(example_inputs) if isinstance(x, torch.SymInt)
-    ]
+    sym_shape_indices: List[int] = []
 
+    compile_run = True
     first_run = True
+    graph_for_symbolic_shape: Callable = None  # type: ignore
 
-    # this is the function we return to Dynamo to run finally
+    # this is the function we return to run finally
     def compiled_graph_wrapper(*args):
+        nonlocal first_run, compile_run
+        nonlocal runtime_shapes_to_compile_flags
+        nonlocal runtime_shapes_to_compiled_entry
+        nonlocal graph_for_symbolic_shape
+        nonlocal sym_shape_indices
 
-        runtime_shapes: Tuple[int,
-                              ...] = tuple(args[i] for i in sym_shape_indices)
+        if compile_run:
+            compile_run = False
 
-        nonlocal first_run
-        nonlocal runtime_shapes_to_compile_flags
-        nonlocal runtime_shapes_to_compiled_graph
+            # this is the first compilation, we will compile a graph with
+            # dynamic shape, as the caller will mark first dimension as dynamic
+
+            if compilation_configs.use_inductor:
+                if is_first_graph:
+                    logger.info("Compiling a graph for general shape")
+                graph_for_symbolic_shape = wrap_inductor(
+                    graph, args, compilation_configs.inductor_compile_config)
+            else:
+                graph_for_symbolic_shape = graph
+
+            # TODO: Dynamo does not pass all dynamic shapes.
+            # Need to investigate why. It works now because all the dynamic
+            # shapes have the same value, and either of them can be used.
+            sym_shape_indices = [
+                i for i, x in enumerate(args) if isinstance(x, torch.SymInt)
+            ]
+
+            return graph(*args)
 
         if first_run:
-            # the first compilation is for profiling, we directly run it
             first_run = False
             return graph_for_symbolic_shape(*args)
 
+        runtime_shapes: Tuple[int,
+                              ...] = tuple(args[i] for i in sym_shape_indices)
+
         if runtime_shapes not in runtime_shapes_to_compile_flags:
             # we haven't seen this shape before
             # query if we need to specialize for this shape
             # we only specialize for the first dimension.
             # TODO: investigate if any model needs to specialize
             # beyond the first dimension
+
             runtime_shapes_to_compile_flags[runtime_shapes] = runtime_shapes[
-                0] in sizes_to_specialize
+                0] in compilation_configs.compile_sizes or \
```

Comment:
This logic is very confusing. What's `runtime_shapes_to_compile_flags`? Seems like it's an array of booleans. And it is not clear what state is kept where, and why there need to be 3 arrays of various shapes

Reply:
It seems like CUDA graph capture sizes are different from compilation sizes, but that should be described a bit better